File: kernels/C++/compositron/src/core/measurement.cpp
#include "compositron/core/measurement.hpp"

#include <ostream>
#include <format>
#include <string>
#include <filesystem>

#include "compositron/importers/meps_dat_importer.hpp"
#include "compositron/importers/slope_n42_importer.hpp"


namespace compositron::core {


namespace measurement {


// class Shape

Shape::Shape(
    size_t dbs, size_t cdbs, size_t pals
) : dbs(dbs), cdbs(cdbs), pals(pals) { }

std::ostream& operator<<(std::ostream& os, Shape const& shape) {
    return os << std::format(
        "Shape(DBS: {}, CDBS: {}, PALS: {})", shape.dbs, shape.cdbs, shape.pals
    );
}


} // namespace measurement


// class Measurement

Measurement::Measurement(const std::string filepath) {
    std::filesystem::path path(filepath);

    if (path.extension() == ".n42") {
        *this = importers::import_SLOPE_n42(filepath);
    } else if (path.extension() == ".dat") {
        *this = importers::import_MePS_dat(filepath);
    }

    filename = path.filename();
    name = filename;
}

measurement::Shape Measurement::shape() const {
    return {
        dbs.size(),
        cdbs.size(),
        pals.size(),
    };
}


} // namespace compositron::core

File: kernels/Python/compositron/src/compositron/core/measurement.py
from pathlib import Path

from ..dbs import DBSpectrum
from ..cdbs import CDBSpectrum
from ..pals import PALSpectrum


class Shape:
    def __init__(self, dbs: int, cdbs: int, pals: int):
        self.dbs = dbs
        self.cdbs = cdbs
        self.pals = pals

    def __str__(self):
        return repr(self)

    def __repr__(self):
        return f"Shape(dbs={self.dbs}, cdbs={self.cdbs}, pals={self.pals})"


class Measurement:
    def __init__(self, filepath: str | None = None):
        self.filename: str | None = filepath
        self.name: str | None = filepath
        self.dbs: dict[str, DBSpectrum] = {}
        self.cdbs: dict[str, CDBSpectrum] = {}
        self.pals: dict[str, PALSpectrum] = {}
        self.metadata: dict[str, str] = {}

        if filepath is not None:
            if Path(filepath).suffix == ".n42":
                from ..importers.slope_n42_importer import import_slope_n42
                m = import_slope_n42(filepath)

            elif Path(filepath).suffix == ".dat":
                from ..importers.meps_dat_importer import import_meps_dat
                m = import_meps_dat(filepath)
            else:
                raise ValueError("Unsupported file format")

            self.dbs = m.dbs
            self.cdbs = m.cdbs
            self.pals = m.pals
            self.metadata = m.metadata

    @property
    def shape(self) -> Shape:
        return Shape(len(self.dbs), len(self.cdbs), len(self.pals))
